src/workerd/api/encoding.c++

```diff
@@ -460,38 +464,584 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
 // =======================================================================================
 // TextEncoder implementation
 
+namespace {
+
+[[maybe_unused]] constexpr inline bool isLeadSurrogate(char16_t c) {
+  return (c & 0xFC00) == 0xD800;
+}
+
+[[maybe_unused]] constexpr inline bool isTrailSurrogate(char16_t c) {
+  return (c & 0xFC00) == 0xDC00;
+}
+
+// Calculate UTF-8 length from UTF-16 with potentially invalid surrogates.
+// Invalid surrogates are counted as U+FFFD (3 bytes in UTF-8).
+// Uses SIMD for valid portions and falls back to scalar for invalid surrogates.
+size_t utf8LengthFromInvalidUtf16(kj::ArrayPtr<const char16_t> input) {
+  size_t inputPos = 0;
+  size_t utf8Length = 0;
+
+  while (inputPos < input.size()) {
+    // Find the next invalid surrogate using SIMD validation
+    auto result =
+        simdutf::validate_utf16_with_errors(input.begin() + inputPos, input.size() - inputPos);
```

Comment:
Nit: you can simplify some of the pointer arithmetic (and make it a bit safer) by slicing the `input` rather than calculating offsets. Unsure of perf difference there so optional.